content/oci/deletableoci.go

```diff
@@ -0,0 +1,216 @@
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package oci provides access to an OCI content store.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+package oci
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/internal/container/set"
+)
+
+// DeletableStore implements `oras.Target`, and represents a content store
+// extended with the delete operation.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+type DeletableStore struct {
+	*Store
+	lock sync.RWMutex
+}
+
+// NewDeletableStore returns a new DeletableStore with context.Background().
+func NewDeletableStore(root string) (*DeletableStore, error) {
+	store, err := NewWithContext(context.Background(), root)
+	if err != nil {
+		return nil, err
+	}
+	ds := &DeletableStore{
+		Store: store,
+	}
+	return ds, nil
+}
+
+// NewDeletableStoreWithContext returns a new DeletableStore.
+func NewDeletableStoreWithContext(ctx context.Context, root string) (*DeletableStore, error) {
+	store, err := NewWithContext(ctx, root)
+	if err != nil {
+		return nil, err
+	}
+	ds := &DeletableStore{
+		Store: store,
+	}
+	return ds, nil
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (ds *DeletableStore) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.Store.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
+func (ds *DeletableStore) Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	return ds.Store.Push(ctx, expected, reader)
+}
+
+// Delete removes the content matching the descriptor from the store.
+func (ds *DeletableStore) Delete(ctx context.Context, target ocispec.Descriptor) error {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	resolvers := ds.tagResolver.Map()
+	untagged := false
+	for reference, desc := range resolvers {
+		if content.Equal(desc, target) {
+			ds.tagResolver.Untag(reference)
+			untagged = true
+		}
+	}
+	if err := ds.graph.Remove(ctx, target); err != nil {
+		return err
+	}
+	if untagged && ds.AutoSaveIndex {
+		err := ds.saveIndex()
+		if err != nil {
+			return err
+		}
+	}
+
+	// we may not need the type assertion, if we change Store.storage from the
+	// read-only interface to a struct pointer.
+	storage := ds.storage.(*Storage)
+	return storage.Delete(ctx, target)
```

Comment:
we may not need type assertion, if we change Store.storage from the read-only interface to a struct pointer. We may discuss on that.

Reply:
We can just change `Store.storage` to `*Storage` type.

Reply:
We can do that, please go ahead.

Reply:
changed `Store.storage` from the read-only interface to a struct pointer, and removed the type assertion.